ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -421,19 +421,58 @@ module Vendor_nvidia = struct
         !Xapi_globs.nvidia_host_driver_file ;
       Xapi_globs.nvidia_default_host_driver_version
 
+  (** Turn an Nvidia version string into a list on numbers *)
+  let nvidia_version version =
+    try
+      Astring.String.filter (( <> ) '+') version
+      |> Astring.String.cuts ~empty:false ~sep:"."
+      |> List.map int_of_string
+    with _ ->
+      error "Not a proper NVidia host driver version: '%s'" version ;
+      failwith (Printf.sprintf "%s is not a version string" version)
+
+  (** [compare_versions] compares two host driver
+  versions represented as int list of possibly different length. The
+  shorter list is conceptually extended with zero entries *)
+  let compare_version v1 v2 =
+    let rec compare v1 v2 =
+      match (v1, v2) with
+      | [], [] ->
+          0
+      | [], 0 :: ys ->
+          compare v1 ys
+      | [], _ ->
+          -1
+      | 0 :: xs, [] ->
+          compare xs v2
+      | _, [] ->
+          1
+      | x :: xs, y :: ys when x < y ->
+          -1
+      | x :: xs, y :: ys when x > y ->
+          1
+      | _ :: xs, _ :: ys ->
+          compare xs ys
+    in
+    compare v1 v2
+
   let host_driver_supports_multi_vgpu ~host_driver_version ~supported_versions =
-    let version = float_of_string host_driver_version in
-    List.exists
-      (fun pattern ->
-        match String.index_opt pattern '+' with
-        | Some i -> (
-          try float_of_string (String.sub pattern 0 i) <= version
-          with _ -> false
-        )
-        | None -> (
-          try float_of_string pattern = version with _ -> false
-        ))
-      supported_versions
+    try
+      let driver = nvidia_version host_driver_version in
+      let supported = List.map nvidia_version supported_versions in
+      let latest =
```

Comment:
I'm a bit confused - do you mean `earliest` here? My understanding is that the earliest version that supports `multi_vgpu` will be marked with a `+`, e.g. 1.2 onwards, or 1.2+

Reply:
Yeah, maybe not the best name. This whole idea is a bit odd and I only kept it for compatibility with what we have. For example: `[1,3+,2+]` is semantically `[1,2+]` - it does not make sense to have more than one entry with `+` but I'm only looking for the first one.